### ISSUE TITLE
EGID usage

### ISSUE THREAD

**Issue by leopripos** [sebas77/Svelto.ECS#26]
I have a case about entity relations using EGID.

1. I have a player entity that has many troop entities.
```csharp
public interface IPlayerTroopComponent
{
         IList <EGID> Troops {get; } // EGID is the troop entity id
}
```
2. Some of my Engine will change the group of troop entities.

The problem is: **Since EGID is a struct, IPlayerTroopComponent will contains inconsistent groupIds.**

What do you think about this?

**Comment by sebas77** [comment]
I firstly think that you shouldn't hold list of entities IDs. Most of the times when you need entities ID you actually need a dedicated group. Can this be the case too?

**Comment by leopripos** [comment]
I've been using groups to group troops by teams.

This is the complete case:
I have Player troops and Non-Player Troops.
- Player troops are owned by players.
- Non-player troops are owned by teams.

Both are team troops.  Then,
- I need to iterate all the troops by certain player. [result: Player Troops]
- I need to repeat all the troops by certain team. [result : Player Troops and Non-Player Troops]

I can't use group because each entity limited in 1 group.

**Comment by sebas77** [comment]
ok I suspected this is the case, I need to think a bit about it. I may have to way to solve it: make the EGID invariant regarding the groupID (I am not sure can be done), or finding another way to query that is not awkward.

In theory you can do this, but I am not sure it's a good idea, as it goes back to the MetaEntity concept:

You create a MetaEntity that has a MetaEntityView like _MetaTroopsPerPlayerEntityView_. This is a new Entity which is in a group that you can query. The MetaEntity is then built using an implementor shared with the normal Entity. This implementor simply returns the real entity EGID. Being the implementor shared between the entity and the metaenttiy, when the entity EGID of the real entity changes, the metaEntity can query it through the metaentityview.

In this way you can effectively regroup entities. This is what the meta entities were for. How does it feel to you?

**Comment by sebas77** [comment]
I think the example is still not very clear to me.

You player and not players (I guess AI?) and they both can have troops. Troops can be split in to teams right? You have engines that iterate over teams regardless if they are player or AI troops. You have also engines that iterate only over the AI troops and only over the player troops. Is all of this correct?

**Comment by sebas77** [comment]
it's theoretically possible to put an entity on several groups, but the awkward part will be the remove entity though.

**Comment by sebas77** [comment]
OK this is my take. To solve your problem I would use two different entitiydescriptor like:

PlayerTroopEntityDescriptor<EnemyTroopEntityView>
NonPlayerTroopEntityDescriptor<PlayerTroopEntityView>

they are gouped by team. but I can query two different entityview...how does it sound? I guess the problem here is that you don't want your engine to know if it's an EnemyTroop or a PlayerTroop. In this case you can still do:

PlayerTroopEntityDescriptor<EnemyTroopEntityView, TroopEntityView>
NonPlayerTroopEntityDescriptor<PlayerTroopEntityView, TroopEntityView>

carefull that this stuff works as long as we are talking about EntityViewClass, if you were using EntitStructs it would have been trickier.

**Comment by leopripos** [comment]
We can go on, I use EntityViewClass.

This is multiplayer game, so player and enemy term just work in client-side, right?
But not in server-side.

Then in server, should I make static player count? like Player1TroopEntityView, Player2TroopEntityView, Player3EntityView?

**Comment by sebas77** [comment]
no if you have more than a player is quite bad

**Comment by sebas77** [comment]
another idea I was pondering is to return an unique ID from the BuildEntity function. I can make this ID never change, but I didn't like the idea, maybe it's the only solution?

**Comment by sebas77** [comment]
I will think about the idea to let the BuildEntity returns an unique ID, remove the BuildEntityInGroup and add a function PutEntityInGroup, let an entity be in multiple groups. I need to check the pros and cons.

**Comment by leopripos** [comment]
Since you said "It's not good for coders writing their own data structures to handle entities", the best to do it which i'm thinking about is using multiple group.

But, can you explain more about "The awkward part of using multiple group will be the remove"?

**Comment by leopripos** [comment]
I it quite hard to explain in english for me, so let me explain by code :)
Here what I want, I belive you can get it.

```csharp
     entityFactory.BuildInGroup<PlayerTroopDecriptror<TroopEntityView>>(id, components, playerID, teamID);
     entityFactory.BuildInGroup<NonPlayerTroopDecriptror<TroopEntityView>>(id, components, teamID);

     var troops= entityDB.QueryGroupedEntityViews<TroopEntityView>(playerID);
     var troops= entityDB.QueryGroupedEntityViews<TroopEntityView>(teamID);
```

Can you?
:)

**Comment by sebas77** [comment]
I understand you want to query the same entity from different groups. I need to think about it, all the consequences and how it works with Entity structs (entity structs cannot be on multiple groups)

**Comment by sebas77** [comment]
BTW what you are doing can be done already, you are effectively building two entities, but sharing the same implementor.

It should work already :)

my thinking is about doing the same with the same entity, but TBH why not doing what you have done?!

**Comment by sebas77** [comment]
Let me be clearer what I meant is that you can build two different entities in two different groups using the same entity descriptor and implementors. Will this be good enough? It means that you can effectively query the same entity data from different groups

**Comment by leopripos** [comment]
If I build two different entities in two different group, but what about destroying entity?

How to identify which entities should be destroyed when destroying entity.

For example, I have X entity, so I should build 2 entities with id 100 and 200.

Then when i want to destory the entity X, how to identify which entities should be destoryed?
Should I make new component to hold that  2 entity ids?

**Comment by leopripos** [comment]
At the end, I will break **"It's not good for coders writing their own data structures to handle entities"** principle.

**Comment by sebas77** [comment]
With EGID you can use the same entity ID in two different groups. Would it
work in this case?

**Comment by leopripos** [comment]
Sorry, i missed that point.

give me time to try, and I'll back.
:)

**Comment by leopripos** [comment]
Sorry, I need more to upgrade my project and ensure everthing well.

Some sort problems :(
1. You removed `QueryIndexableEntityViews<T>(groupId);`
2. I need create EGID when `QueryEntityView<T>(EGID);`

**Comment by leopripos** [comment]
> The first function was redundant you can query the entity directly without
passing through the dictionary right?

Yes, not big problem, but make me change how I write my code :)

About the second, I think we have painful part when working EGID.

Since we need create EGID when querying entity, `QueryEntityView<T>(EGID)`, we need to pass groupId to every engines even though we have unique entityId. 
This is happen especially when working with Networking.

The case of networking:
I have 2 project client and server.
Client need to upgrade level specific troop, and send upgrade command to server with parameters TroopID.
Then in server, we need to identify the groupId of that TroopID. It the hard part, right?

What do you think?

**Comment by sebas77** [comment]
When I changed it I knew that it would have been less convenient, but it's
the only way to be able to use same ids on different groups. Is the group
different between server and client?

**Comment by leopripos** [comment]
Sorry not metion it. Yes, It is different.

But, I just found the solution. Because of all my entities have unique Id, I just need build the entity in `ExclusiveGroups.StandardEntity` group. 
It will make me just need to access using ExclusiveGroups.StandardEntity.

Is it ok?

**Comment by sebas77** [comment]
You can create an EGID with ID only it will use the standard group
automatically

**Comment by leopripos** [comment]
Yeah  :)

But will it cause performance issue?

When I build 3 entity using same entityID, will it build 3 EntityViews? Will it call reflection 3 times?

**Comment by sebas77** [comment]
Reflection is called once per entity build per entity descriptor. I may
optimize it to exploit inheritance and I have other optimizations in future
to build an object per entity and not entity view. All that said you
shouldn't use build entity with entity view in performance critical code
path

**Comment by sebas77** [comment]
To be clear if you called build entity twice with the same entity
descriptor the reflection is used once and cached from that moment on

**Comment by leopripos** [comment]
Nice one about cached reflection.

About our solution above, "Building multiple entity using same EntityId".
I just check it. When we build 3 entity using same entityId, it will build 3 EntityViews.
It will causing problem when engine using `EntityDB.QueryEntityViews<T>()`. It will return 3 EntityView with shared components.

**Comment by sebas77** [comment]
I know that's a future optimization I am still working on and it will
change a bit how to declare entity views

**Comment by sebas77** [comment]
let me know if the solution worked out for you in the end or you found it too awkward ;)

**Comment by leopripos** [comment]
Yes, I think little awkward. :)

I changed the internal implementation of `EntityDB.QueryEntityViews<T>()` be querying from `ExclusiveGroups.StandardEntity`.
Then All entities should be built in `ExclusiveGroups.StandardEntity` group.
:D

Btw, I just see coming update announcement. Are there updates related to this?

**Comment by sebas77** [comment]
QueryEntityViews -> Queries from the standard group
QueryGroupedEntityViews -> Queries from whatever group
TryQueryEntityView -> gets the specific entity from whatever group using the EGID

**Comment by leopripos** [comment]
> About our solution above, "Building multiple entity using same EntityId".
I just checked it. When we build 3 entity using same entityId, it will build 3 EntityViews.
It will cause problem when engine using `EntityDB.QueryEntityViews<T>()`. It will return 3 EntityView with shared components.

Here the problem.

QueryEntityViews not queries by group. It queries from GlobalEntityViews.

https://github.com/sebas77/Svelto.ECS/blob/fe0a777475a3f2b7570539e12b96fc16d74c40c4/Svelto.ECS/EntityViewsDB.cs#L24

**Comment by sebas77** [comment]
why aren't you using QueryGroupedEntityViews though?

**Comment by sebas77** [comment]
Regardless your problem, the _globalEntityViewsDB is something that I am torn if removing or not. This change the design a lot, as currently you can create engines that handle grouped entity views and engines that handle all the entityviews of a given type regardless which group they are in. If I remove the globalEntityViewsDB, then you must always know in which groups your entities are and you can't design anymore engines that are not aware of groups.

At the same time if groups are not needed, the StandardGroup can be used as "global pool", while if groups are needed the engines must be aware of the group.

However I think this is not a good idea, as if I have grouped entities (like enemy entities) with HealthEntityView, I cannot write anymore an engine that handles all the HealthEntityView regardless the group the entities are in.

**Comment by leopripos** [comment]
Just for making consistent code.

Since I built multiple entities using same EntityId which causing problem above, `QueryEntityViews<T>()` become not reliable API.
Then, I don't like "**Sometimes we use `QueryEntityViews<T>()` to query all EntityViews, somitimes we use `QueryGroupedEntityViews<T>(ExclusiveGroups.StandardEntity)` to query all EntityViews** " 

So I changed it, and make new constaint that all entities should be built in `ExclusiveGroups.StandardEntity` group.

**Comment by sebas77** [comment]
no, read above, that is not as it works. There is currently a very good reason why the global entity DB must be used separated by the groups. Unless I found another solution, it must stay as it is.

**Comment by sebas77** [comment]
if I remove the global pool, then the new design really must be to build multiple entities to use group. 
I.E.:

* BuildEntityInGroup<SoldierEntityDescriptor>(ID);  
* BuildEntityInGroup<SoldierEntityDescriptor>(ID, teamID);
* BuildEntityInGroup<SoldierEntityDescriptor>(ID, playerID);

in this way I build the entity three times because I want it in the global standard pool, grouped per team and grouped per player.

I mean, if we agree it's not too bad, it may be a solution

**Comment by leopripos** [comment]
I agree about `QueryEntityViews()` and `QueryGroupedEntityViews(ExclusiveGroups.StandardEntity)` are two different things. 

But, what I'm thinking now, with current Svelto.ECS, the main problem is about how we build entities, not querying entities. 
And I also believe (by intuition), that you are also thinking about it. XD

So, to make my project easier to upgrade to new version of Svelto.ECS, I chose using modified `QueryEntityView<T>()` to query all EntityViews instead of  using `QueryGroupedEntityViews<T>(ExclusiveGroups.StandardEntity)`. I think it is not bad as long I use it to query all EntityViews.

What is the easier part? The easier part is I just need to focus at my factories, not factories and  which engines using QueryGroupedEntity<T>(ExclusiveGroups.StandardEntity).

does it make sense?
:)

**Comment by sebas77** [comment]
If you agree that what I wrote above makes sense, I'll take that direction, but you shouldn't change the framework code unless you know that I could break it :)

**Comment by leopripos** [comment]
Yeah, my bad. :)

What do you think like this?
```csharp
void BuildEntityInGroup(int id, params int[] groups);
```

**Comment by leopripos** [comment]
Or may be better just using `BuildEntity`
```csharp
void BuildEntity(int id, params int[] groups);
```

**Comment by sebas77** [comment]
you want me to hide the fact that the entity is actually built multiple times? It may be an idea

**Comment by leopripos** [comment]
Yeah it is much cleaner.

Then, when destroying, will we still need EGID? or just EntityID?

**Comment by leopripos** [comment]
Since you already have live product which not using unique Id, I think it will be much harder.

Sorry, causing problems for you.
:D

**Comment by sebas77** [comment]
it's not just that, if the user decides to use groups, it must mind them. If decide to use more than one group, it must mind them. This means that the user must in any moment know in which groups the entities are. If not, you can easily get lost and having entities everywhere! We must be very careful and keep it under control.

**Comment by Darelbi** [comment]
@leopripos If you need just to iterate combinations of 2 groups, and you want to avoid the overhead of copying all elements of 2 groups into a new array, why you just don't iterate groups in sequence? 

The most silly implementation I can come with is

    public class CombinedGroupIterator<T> where T : EntityView, new()
    {
        private readonly int[] groupsToCombine;

        public CombinedGroupIterator( int[] groupsToCombine)
        {
            this.groupsToCombine = groupsToCombine;
        }

        public IEnumerator< T> GetEnumerator( IEntityViewsDB entityViewsDB)
        {
            foreach(var group in groupsToCombine)
            {
                int count = 0;
                foreach(var view in entityViewsDB.QueryGroupedEntityViewsAsArray<T>( group, out count))
                {
                    yield return view;
                }
            }
        }

    }

Usage in your engine:

        public class MyEngine: IQueryingEntityViewEngine
    {
        private readonly CombinedGroupIterator<MyEntityView> combinedGroup = null;

        public MyEngine()
        {
            int group1 = 3;
            int group3 = 4;
            int group5 = 7;

            combinedGroup = new CombinedGroupIterator< MyEntityView>( 
                new int[] { group1, group3, group5 }); 
        }

        public IEntityViewsDB entityViewsDB { set; private get; }

        public void Ready() {  }

        public void DoSomething()
        {
            var enumerator = combinedGroup.GetEnumerator( entityViewsDB); // 1 allocation

            while (enumerator.MoveNext())
            {
                MyEntityView view = enumerator.Current;
            }
        }
    }

Of course Getting a enumerator implies doing 1 allocation, but you can workaround that, this is the slightly better version:

    public class BetterCombinedGroupIterator<T> where T : EntityView, new()
    {
        private readonly int[] groupsToCombine;

        public BetterCombinedGroupIterator( int[] groupsToCombine)
        {
            this.groupsToCombine = groupsToCombine;
        }

        public IEnumerator< T> GetEnumerator( IEntityViewsDB entityViewsDB)
        {
            while (true)
            {
                foreach (var group in groupsToCombine)
                {
                    int count = 0;
                    foreach (var view in entityViewsDB.QueryGroupedEntityViewsAsArray<T>(group, out count))
                    {
                        yield return view;
                    }
                }

                yield return null; // this notify the end of a cycle of iterations
            }
        }

    }

This can still be used, but have the advantage of performing no more allocations at runtime:

    public class MyEngine: IQueryingEntityViewEngine
    {
        private readonly BetterCombinedGroupIterator< MyEntityView> combinedGroup = null;
        private IEnumerator< MyEntityView> viewCollection = null;

        public MyEngine()
        {
            int group1 = 3;
            int group3 = 4;
            int group5 = 7;

            combinedGroup = new BetterCombinedGroupIterator< MyEntityView>( 
                new int[] { group1, group3, group5 }); 
        }

        public IEntityViewsDB entityViewsDB { set; private get; }

        public void Ready()
        {
            viewCollection = combinedGroup.GetEnumerator( entityViewsDB);
        }

        public void DoSomething()
        {
            MyEntityView currentView = null;

            while( viewCollection.MoveNext() && (currentView = viewCollection.Current) != null)
            {
                // do what ya want with currentview.. No allocations at all!
            }
        }
    }

And I think with minor effort you can create a class from which your engine Inherit to further reduce the boilerplate code.

**Comment by sebas77** [comment]
it's an interesting approach, I think it can be improved too. Something to keep in mind.

**Comment by sebas77** [comment]
@leopripos another way you can explore is actually to build multiple EnginesRoot. There is no limitation to the number of EnginesRoot you can build. Multiple engines root are useful to encapsulate logic even more.

**Comment by leopripos** [comment]
@Darelbi Looks interesting, Can you explain more how to handle this case?

> entityFactory.BuildInGroup<PlayerTroopDescriptror<TroopEntityView>>(id, components, playerID, teamID);
>     entityFactory.BuildInGroup<NonPlayerTroopDescriptror<TroopEntityView>>(id, components, teamID);

 >    var troops= entityDB.QueryGroupedEntityViews<TroopEntityView>(playerID); // return player troops
>     var troops= entityDB.QueryGroupedEntityViews<TroopEntityView>(teamID); // return player and non-player troops

**Comment by Darelbi** [comment]
It is Just the same. Also Player and No player descriptors should share
some views. And you no longer have to build the same entity in multiple
groups.

    BuildInGroup (playerId);//Just player id
    BuildInGroup (teamid);

    var combinedTeam = new Grouped( new []{ teamid, playerId });
    var justPlayer = new Grouped( new []{ playerId});
    var justTeam = new Grouped( new[]{ teamid});

    while( combinedTeam.MoveNext())
    {
         var view = combinedTeam.Current;
         // All' player and non player troops
    }

    while( justTeam.MoveNext())
    {
         var view = justTeam.Current;
         // All troops of that team not owned by any player
    }


    while( justPlayer.MoveNext())
    {
         var view = combinedTeam.Current;
         // All' player troops
    }

Note you dont Need combined groups for groups of 1 group only. So the
"Just" stuff already workshop out of the box in svelto.


Il mercoledì 18 aprile 2018, Leo Pripos Marbun <notifications@github.com>
ha scritto:

**Comment by sebas77** [comment]
If you use wisely it's actually suggested. For example I have seen cases of
UI written with their own encapsulated engines. When the division is clear
separate engines root are beneficial
